[WebGPU] RenderBundleEncoder::storeVertexBufferCountsForValidation() release asserts on destroyed index buffers

mwyrzykowski opened the pull request:
#### e9f9b68e0723c53592eadd320e0256bcae6319ef
<pre>
[WebGPU] RenderBundleEncoder::storeVertexBufferCountsForValidation() release asserts on destroyed index buffers
<a href="https://bugs.webkit.org/show_bug.cgi?id=291180">https://bugs.webkit.org/show_bug.cgi?id=291180</a>
<a href="https://rdar.apple.com/148704622">rdar://148704622</a>

Reviewed by Cameron McCormack.

There were three bugs here:

  1. An index buffer destroyed between the calls of encoder.setIndexBuffer() and
     encoder.drawIndexed() would trigger a release assertion. Mirror what RenderPassEncoder
     does and skip the draw call in this case.
  2. this validation container is not needed for record+replay path,
     as that goes through regular render pass validation and there
     exists no ICBs to mutate
  3. When a RenderBundle was composed of multiple ICBs, m_minVertexCountForDrawCommand
     was never cleared leading to redundant validation passes. Correct this by clearing
     as we do the other containers.

* LayoutTests/fast/webgpu/nocrash/fuzz-291180-expected.txt: Added.
* LayoutTests/fast/webgpu/nocrash/fuzz-291180.html: Added.
Add regression test.

* Source/WebGPU/WebGPU/RenderBundleEncoder.mm:
(WebGPU::RenderBundleEncoder::storeVertexBufferCountsForValidation):
(WebGPU::RenderBundleEncoder::endCurrentICB):

Canonical link: <a href="https://commits.webkit.org/293423@main">https://commits.webkit.org/293423@main</a>
</pre>

heycam commented on Source/WebGPU/WebGPU/RenderBundleEncoder.mm:
```diff
@@ -749,7 +753,8 @@ - (instancetype)initWithICB:(id<MTLIndirectCommandBuffer>)icb containerBuffer:(i
     auto indexCountTimesSizeInBytes = checkedProduct<size_t>(indexCount, indexSizeInBytes);
     if (id<MTLIndirectRenderCommand> icbCommand = currentRenderCommand()) {
-        storeVertexBufferCountsForValidation(indexCount, instanceCount, indexBufferOffsetInBytes / indexSizeInBytes, baseVertex, firstInstance, m_indexType, indexBufferOffsetInBytes);
+        if (!storeVertexBufferCountsForValidation(indexCount, instanceCount, indexBufferOffsetInBytes / indexSizeInBytes, baseVertex, firstInstance, m_indexType, indexBufferOffsetInBytes))
+            return finalizeRenderCommand();
```
Do we need to makeInvalid in here?

mwyrzykowski replied:
I don’t think so because this is due to a very confusing (to me at least) part of the specification where ‘destroyed objects should continue to function like normal objects’, until they are submitted as part of a GPUCommandBuffer

mwyrzykowski replied:
See https://gpuweb.github.io/gpuweb/explainer/#early-destroy

> Unlike invalidity, destroyed-ness can change after creation, is not contagious, and is validated only when work is actually submitted (e.g. queue.writeBuffer() or queue.submit()), not when creating dependent objects (like command encoders, see above).

we had initially not assumed this 2 years ago and all the related tests failed

mwyrzykowski replied:
Likely the motivation is while the index buffer is used as part of the draw call with Metal, on another backend, it is only used in the setIndexBuffer call

mwyrzykowski replied:
We handle the make the command buffer commit invalid via setCommandEncoder on the index buffer

mwyrzykowski committed:
[WebGPU] RenderBundleEncoder::storeVertexBufferCountsForValidation() release asserts on destroyed index buffers
https://bugs.webkit.org/show_bug.cgi?id=291180
rdar://148704622

Reviewed by Cameron McCormack.

There were three bugs here:

  1. An index buffer destroyed between the calls of encoder.setIndexBuffer() and
     encoder.drawIndexed() would trigger a release assertion. Mirror what RenderPassEncoder
     does and skip the draw call in this case.
  2. this validation container is not needed for record+replay path,
     as that goes through regular render pass validation and there
     exists no ICBs to mutate
  3. When a RenderBundle was composed of multiple ICBs, m_minVertexCountForDrawCommand
     was never cleared leading to redundant validation passes. Correct this by clearing
     as we do the other containers.

* LayoutTests/fast/webgpu/nocrash/fuzz-291180-expected.txt: Added.
* LayoutTests/fast/webgpu/nocrash/fuzz-291180.html: Added.
Add regression test.

* Source/WebGPU/WebGPU/RenderBundleEncoder.mm:
(WebGPU::RenderBundleEncoder::storeVertexBufferCountsForValidation):
(WebGPU::RenderBundleEncoder::endCurrentICB):

Canonical link: https://commits.webkit.org/293423@main
